### PR TITLE
Fix the edge case when handling non numeric values of double type in delta stats

### PR DESCRIPTION
**Pull request by emilie-wang**
## What is the purpose of the pull request

The pr aims to fix: https://github.com/apache/incubator-xtable/issues/524.
When reading the delta snapshot and load the information into Delta object AddFile, the non-numeric values of float or double type (example, "NaN", "-Infinity") from col stats become string type. These special values need special handling and this pr used the same idea how delta new API handled this: https://github.com/delta-io/delta/blob/master/kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/data/DefaultJsonRow.java#L210
## Brief change log
- Add logic to handle special String value ("NaN", "-Infinity", "Infinity") when column type is Double or Float during translating Delta column stat to InternalType 
- Add unit test

## Verify this pull request

This change added tests and can be verified as follows:

- Added unit test for this edge case

### CODE REVIEW

**Comment by emilie-wang** [comment]
Hi, Can I have your help to rerun the Ci build pipeline. I couldn't replay the same error locally:
```
Error:  org.apache.xtable.hudi.extensions.TestAddFieldIdsClientInitCallback.existingTable -- Time elapsed: 0.219 s <<< ERROR!
java.lang.NullPointerException
```

File: xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java [hunk 1]
```diff
@@ -194,9 +194,35 @@ public static Object convertFromDeltaPartitionValue(
 
   private static Object castObjectToInternalType(Object value, InternalType valueType) {
     switch (valueType) {
+      case DOUBLE:
+        if (value instanceof String)
+          switch (value.toString()) {
+            case "NaN":
```
Comment (the-other-tim-brown) [inline xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java:200]: Is there anywhere we can import these strings as constants from Delta? if not, can you define some internally for us?

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java [hunk 2]
```diff
@@ -82,6 +82,15 @@ void parseWrongDateTime() throws ParseException {
     assertThrows(ParseException.class, () -> strictDateFormat.parse(wrongDateTime));
   }
 
+  @ParameterizedTest
+  @MethodSource("nonNumericValuesForColStats")
+  public void formattedDifferentNonNumericValuesFromDeltaColumnStat(
```
Comment (the-other-tim-brown) [inline xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java:87]: nitpick: `public` can be removed

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java [hunk 3]
```diff
@@ -82,6 +82,15 @@ void parseWrongDateTime() throws ParseException {
...
+      Object fieldValue, InternalSchema fieldSchema, Object expectedDeltaValue) {
+    Object internalRepresentation =
+        DeltaValueConverter.convertFromDeltaColumnStatValue(fieldValue, fieldSchema);
+    assertEquals(internalRepresentation, expectedDeltaValue);
```
Comment (the-other-tim-brown) [inline xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java:91]: Do we also need to test transforming the internal representation into the delta stats?

**Comment by the-other-tim-brown** [comment]
> Hi, Can I have your help to rerun the Ci build pipeline. I couldn't replay the same error locally:
> 
> ```
> Error:  org.apache.xtable.hudi.extensions.TestAddFieldIdsClientInitCallback.existingTable -- Time elapsed: 0.219 s <<< ERROR!
> java.lang.NullPointerException
> ```

@emilie-wang I re-ran it for you and everything is passing. Just had some minor comments and questions. Thank you for the quick fix!

File: xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java (hunk 1 above)
Comment (emilie-wang) [reply xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java:200]: After searching there was no constants defined in Delta and will define them internally. 
How about defining them inside DeltaValueConverter class like
```
public class DeltaValueConverter {
  private static final String DATE_FORMAT_STR = "yyyy-MM-dd HH:mm:ss";
  private static final TimeZone TIME_ZONE = TimeZone.getTimeZone("UTC");
  protected static final String NAN_VALUE = "NaN";
  protected static final String INFINITY_VALUE = "Infinity";
...
```

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java (hunk 2 above)
Comment (emilie-wang) [reply xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java:87]: For info, I use `public` by following these tests: https://github.com/apache/incubator-xtable/blob/main/xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java#L44
but happy to remove all of them

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java (hunk 3 above)
Comment (emilie-wang) [reply xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java:91]: Good question, we suppose the internal representation always has correct field value which corresponding to field schema, and existing tests have already covered the case.

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java (hunk 2 above)
Comment (the-other-tim-brown) [reply xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java:87]: We have a linter that flags this stuff which is why I selfishly flagged this for my own internal cherry-pick later :) I am fine with just doing this on new tests for now. Maybe later we can integrate a linter into the the ASF toolkit as well

File: xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java (hunk 1 above)
Comment (the-other-tim-brown) [reply xtable-core/src/main/java/org/apache/xtable/delta/DeltaValueConverter.java]: Thanks, looks good to me!

**Comment by the-other-tim-brown** [comment]
@emilie-wang one final request is for you to squash down to a single commit for me so I can approve and merge this into the main branch. Thanks again for the contribution!

File: xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java (hunk 2 above)
Comment (emilie-wang) [reply xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaValueConverter.java]: Sure. Removed.

**Comment by the-other-tim-brown** [comment]
> Hi @the-other-tim-brown, I am back from vacation and just updated my PR based on your comments. Thank you for your review!

@emilie-wang hope you had a good vacation! Can you squash this all down to 1 commit for me? I've used a command like `git reset --soft $(git merge-base main HEAD)` in the past to do this

**Review by the-other-tim-brown — APPROVED** [review]

**Comment by emilie-wang** [comment]
@the-other-tim-brown sorry I forgot this mesage and just squashed down to 1 single commit. Thank you for the quick review again.
